src/coreclr/jit/gentree.h

```diff
@@ -6217,7 +6227,20 @@ class IntrinsicNodeBuilder final
 struct GenTreeJitIntrinsic : public GenTreeMultiOp
 {
 protected:
-    GenTree*           gtInlineOperands[2];
+    union
+    {
+        GenTree* gtInlineOperands[2];
```

Comment:
I don't think I follow what gtInlineOperands (and the logic inside SetMethodHandle does, can you explain?

Reply:
`gtInlineOperands` is an existing field that allows `GenTreeMultiOp` to avoid allocating for the common case where there is 2 or less operands and instead allows us to carry the fields within the main allocation instead.

In order to facilitate rewriting the hwintrinsic back to a call, however, we need to be able to track the `CORINFO_METHOD_HANDLE` and we don't have enough free space to do that. Adding the field directly ends up making the node larger than `TREE_NODE_SZ_LARGE` as well, so to avoid pessimizing the rest of the JIT I made this into a union. `SetMethodHandle` will then force an allocation if `gtInlineOperands` was being used and set a flag that indicates that `gtMethodHandle` is set instead.

It's worth noting the actual `gtInlineOperands` field isn't ever read directly either, it's address is just handed down to `GenTreeMultiOp` as part of construction or `ResetOperandArray` call. So this ends up working fairly well and ensures that the allocation is "pay for play" and only used when actually necessary (we have an intrinsic with 1 or 2 operands that requires a constant and needs to be carried through to a later phase).

Reply:
-- Notably the reason we don't have enough free space is primarily because of padding bytes caused by inheritance. There's a few places where nodes are wasting 4-7 bytes of space to maintain 8-byte alignment and that repeats several times.

We could avoid tricks like the one being employed here if we had a better mechanism for avoiding such wasted padding for derived node kinds. But that's a much more complex and independent work item.

Reply:
Ah I didn't notice it was a pre-existing field - I though you added it in this PR 🙂